llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,282 @@
+                                   const TypesMap2 &Types) {
+  dbgs() << "CheckInstUsesOptionalDeviceFeature: " << I << "\n";
+  const Type *ReturnType = I.getType();
+  dbgs() << "Check ReturnType: " << *I.getType() << "\n";
+  auto it = CheckTypeContainsOptionalType(ReturnType, Types);
+  if (it != Types.end()) {
+    dbgs() << "Exit\n";
+    return it;
+  }
+
+  dbgs() << "Return Type doesn't contain searched type\n";
+  dbgs() << "Check Operands. NumOperands: " << I.getNumOperands() << "\n";
+  std::string type;
+  int i = 0;
+  for (auto OpIt = I.op_begin(); OpIt != I.op_end(); ++OpIt, ++i) {
+    dbgs() << "Check Operand: " << i << ", : " << **OpIt << "\n";
+    it = CheckTypeContainsOptionalType((*OpIt)->getType(), Types);
+    if (it != Types.end()) {
+      dbgs() << "Searched type found: " << it->first << "\n";
+      return it;
+    }
+  }
+
+  dbgs() << "No type is found\n";
+  return Types.end();
+}
+
+class Graph {
+public:
+  struct Node {
+    Function *F = nullptr;
+    uint32_t count = 0;
+    std::vector<int> edges;
+    std::vector<int> reversedEdges;
+    SmallSet<int, 10> aspects;
+
+    Node(Function &F) : F(&F) {}
+  };
+
+  void addEdge(Function *from, Function *to) {
+    if (functionIndex.count(from) == 0 || functionIndex.count(to) == 0) {
+      assert(false && "couldn't find entries for functions");
+    }
+
+    int indexFrom = functionIndex[from];
+    int indexTo = functionIndex[to];
+    nodes[indexFrom].edges.push_back(indexTo);
+    nodes[indexTo].reversedEdges.push_back(indexFrom);
+  }
+
+  std::vector<Node> nodes;
+  std::unordered_map<Function *, int> functionIndex;
+};
+
+void dfs2(const Graph &G, int NodeIndex, std::vector<int8_t> &Visited,
+          std::vector<int> &Topsort) {
+  if (Visited[NodeIndex])
+    return;
+
+  Visited[NodeIndex] = 1;
+  for (int edge : G.nodes[NodeIndex].reversedEdges) {
+    dfs2(G, edge, Visited, Topsort);
+  }
+
+  Topsort.push_back(NodeIndex);
+}
+
+std::vector<int> dfs(const Graph &G, const std::vector<Function *> Kernels) {
+  std::vector<int> topsort;
+  std::vector<int8_t> visited(G.nodes.size(), 0);
+  for (Function *F : Kernels) {
+    int index = G.functionIndex.at(F);
+    dfs2(G, index, visited, topsort);
+  }
+
+  return topsort;
+}
+
+template <typename T> void PrintMap(const T &m) {
+  dbgs() << "Map: size: " << m.size() << "\n";
+  for (const auto it : m) {
+    dbgs() << "key: " << it.first << "\n";
+  }
+}
+
+PreservedAnalyses PropagateAspectUsagePass::run(Module &M,
+                                                ModuleAnalysisManager &MAM) {
+  auto TypesWithAspects = GetTypesWithAspectsFromMetadata(M);
+  PrintMap(TypesWithAspects);
+
+  Graph g;
+  for (Function &F : M.functions()) {
+    auto CC = F.getCallingConv();
+    if (CC != CallingConv::SPIR_FUNC && CC != CallingConv::SPIR_KERNEL)
+      continue;
+
+    g.nodes.push_back(Graph::Node(F));
+    g.functionIndex[&F] = g.nodes.size() - 1;
+  }
```

Comment:
I feel like this graph traversal can be implemented in a bit more simple way, without introducing extra classes/structures:
```c++

using FunctionToAspectsMapTy = DenseMap<Function *, SmallSet<int, 4>>;
FunctionToAspectsMapTy FunctionToAspectsMap;
using CallGraphTy = DenseMap<Function *, SmallPtrSet<Function *, 8>>;
CallGraphTy CallGraph;
SmallVector<Function *, 16> Kernels;

for (Function &F : M.functions()) {
  if (/* not spir func or spir kernel */)
    continue;
  if (/* is kernel */)
    Kernels.push_back(&F);
  
  for (const auto &I : F.instructions()) {
    using RangeTy = /* ... */;
    RangeTy Range = getAspectsUsedByInstruction(I, TypesWithAspects);
    FunctionToAspectsMap[F].insert(Range.first, Range.second);
    // while we are here, let's update call graph
    if (auto *CI = dyn_cast<CallInst>(I)) {
      if (!CI->isIndirectCall())
        CallGraph[F].insert(CI->getCalledFunction());
    }
  }
}

SmallPtrSet<Function *, 16> Visited;
for (Function &F : Kernels) {
  propagateAspects(F, CallGraph, FunctionsToAspectsMap, Visited);
}

void propagateAspects(Function *F, const CallGraphTy &CallGraph, FunctionToAspectsMapTy &AspectsMap, SmallPtrSet<Function *, 16> Visited) {
  SmallSet<int, 8> LocalAspects;
  for (Function *Callee : CallGraph[F]) {
    if (!Visited.contains(Callee)) {
      Visited.insert(Callee);
      propagateAspectsImpl(Callee, CallGraph, AspectsMap, Visited);
    }
    LocalAspects.insert(AspectsMap[Callee].begin(), AspectsMap[Callee].end());
  }

  // update aspects map to re-use it next time
  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
  
  // set metadata to the function
  // ...
  F->setMetadata(/* ... */);
}
```